File: androlua/src/main/java/cn/vove7/androlua/RemoteDebugServer.java
package cn.vove7.androlua;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.net.ServerSocket;
import java.net.Socket;

import cn.vove7.androlua.luautils.LuaPrinter;

class RemoteDebugServer extends Thread {
    public boolean stopped;
    private Context context;
    private LuaHelper luaHelper;

    public RemoteDebugServer(Context context) {
        this.context = context;
        this.luaHelper = new LuaHelper(context);
    }

    private final static int LISTEN_PORT = 3333;

    @Override
    public void run() {
        stopped = false;
        try (ServerSocket server = new ServerSocket(LISTEN_PORT)) {
            show("Server started on port " + LISTEN_PORT);
            Socket client = server.accept();
            DataInputStream inputStream = new DataInputStream(new BufferedInputStream(client.getInputStream()));
            final DataOutputStream outputStream = new DataOutputStream(new BufferedOutputStream(client.getOutputStream()));

            final StringBuffer buffer = new StringBuffer();
            LuaPrinter.OnPrint print = new LuaPrinter.OnPrint() {
                @Override
                public void onPrint(int l, String output) {
                    buffer.append(output);
                }
            };
            luaHelper.regPrint(print);
            while (!stopped) {
                String data = inputStream.readUTF();
                Log.d("Vove :", "run  ----> " + data);
                luaHelper.safeEvalLua("require 'import'\n" + data);
                outputStream.writeUTF(buffer.toString());
                buffer.setLength(0);
                outputStream.flush();
            }
            show("RemoteDebug finished！");
        } catch (Exception e) {
            e.printStackTrace();
            show(e.getMessage());
        }
    }

    private void show(final String s) {
        Log.d("Vove :", "show  ----> " + s);
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(context, s, Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: androlua/src/main/java/cn/vove7/androlua/RemoteDebugServer.kt
package cn.vove7.androlua

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import cn.vove7.androlua.luautils.LuaPrinter
import cn.vove7.appbus.AppBus
import cn.vove7.datamanager.parse.model.Action
import cn.vove7.datamanager.parse.model.Param
import cn.vove7.vtp.log.Vog
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.net.ServerSocket

internal class RemoteDebugServer(private val context: Context) : Thread() {
    var stopped: Boolean = false


    override fun run() {
        stopped = false
        var print: LuaPrinter.OnPrint? = null
        try {
            ServerSocket(LISTEN_PORT).use { server ->
                show("Debug Server started on port $LISTEN_PORT")
                val client = server.accept()
                val inputStream = DataInputStream(BufferedInputStream(client.getInputStream()))
                val outputStream = DataOutputStream(BufferedOutputStream(client.getOutputStream()))

                val buffer = StringBuffer()
                print = object : LuaPrinter.OnPrint {
                    override fun onPrint(l: Int, output: String) {
                        buffer.append(output)
                    }
                }
                LuaHelper.regPrint(print!!)
                while (!stopped) {
                    val data = inputStream.readUTF()
                    Vog.d(this, "run  ----> $data")
//                    luaHelper.safeEvalLua("require 'bridges'\n$data")
                    post(data)
                    outputStream.writeUTF(buffer.toString())
                    buffer.setLength(0)
                    outputStream.flush()
                }
                show("RemoteDebug finished！")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            show("${e.message}")
        } finally {
            LuaHelper.unRegPrint(print!!)
        }
    }

    fun post(src: String) {
        val ac = Action(src)
        ac.param = Param()
        AppBus.post(ac)
    }

    private fun show(s: String) {
        Vog.d(this, "show  ----> $s")
        Handler(Looper.getMainLooper()).post {
            Toast.makeText(context, s, Toast.LENGTH_SHORT).show()
        }
    }

    companion object {

        private const val LISTEN_PORT = 3333
    }
}